### ISSUE TITLE
[mbr] hot reload doesn't work on lambda bodies

### ISSUE THREAD

**Issue by SteveSandersonMS** [dotnet/runtime#50249]
### Description

The hot reload feature seems to work a lot better on `Index.razor` than on certain other files. I'm not sure why the behavior differs.

Steps:

* Use latest preview 3 SDK
* `dotnet new blazorwasm -o MyApp`
* `cd MyApp`
* `dotnet watch`
* When the app is running, modify some text in `Pages/Index.razor` - see it appears correctly in the browser
* Now try a similar edit inside `Shared/NavMenu.razor`, for example change the text `Home` to `Home2`.
  * Expected: change is applied
  * Actual: runtime crashes

![image](https://user-images.githubusercontent.com/1101362/112522487-9a1a7200-8d95-11eb-883a-ef1a265afc74.png)

### Configuration

This is with SDK 6.0.100-preview.3.21173.6, with `Microsoft.NETCoreSdk.BundledVersions.props` patched to use the runtime version 6.0.0-preview.4.21174.4.

cc @lambdageek

**Comment by SteveSandersonMS** [comment]
@lambdageek I've no idea if this is also addressed by https://github.com/dotnet/runtime/pull/50248 but thought it best to report it rather than waiting to see.

**Comment by lambdageek** [comment]
@SteveSandersonMS doesn't look like it's the same thing.

@SteveSandersonMS when reporting these, one thing that would be super helpful:
1. Save the baseline .dll
2. Open a JS console and put a breakpoint in `aspnetcore-browser-refresh.js` in the `applyBlazorDeltas` function inside the `foreach`.
3. When it hits the breakpoint, in the console display the values of `d.metadataDelta` and `d.ilDelta` and save them. (please make sure that the JS console isn't eliding the middle with "...")
4. hit Continue in the JS debugger and repeat for any other deltas.

**Comment by lambdageek** [comment]
The body of the `NavLink` component is a `RenderFragment` lambda, and changing `"Home" => "Home2"` is updating the lambda body.  It's emitting new fields and methods on the lambda, which Mono MBR doesn't support yet.  (ie this is a rude edit - but we need https://github.com/dotnet/roslyn/issues/49010 in order to signal it)

**Comment by SteveSandersonMS** [comment]
Thanks for clarifying! That makes sense. Glad to hear there’s a plan to support this eventually too. 

RenderFragments are extremely common in Blazor, so it would definitely be super valuable to support editing them.

**Comment by lambdageek** [comment]
(Notes to myself)

Static lambdas generate an interesting closure that tries to only create the delegate for the function just once:
```csharp
public class D {
    public void Foo (Action<D> doer) {
        doer (this);
    }
    public void Bar (int j, int k) {}
}
public class C {
    public int f;
    public void M(D d) {
       int j = 42;
       d.Foo ( (dm1) => dm1.Bar(42,42));
       d.Foo ( (d0) => { HalperClosed(d0); });
   }
}
```

generates

```csharp
```[Serializable]
    [CompilerGenerated]
    private sealed class <>c
    {
        public static readonly <>c <>9 = new <>c();

        public static Action<D> <>9__1_0;

        public static Action<D> <>9__1_1;

        internal void <M>b__1_0(D dm1)
        {
            dm1.Bar(42, 42);
        }

        internal void <M>b__1_1(D d0)
        {
            <M>g__HalperClosed|1_4(d0);
        }
    }

    public int f;

    public void M(D d)
    {
        d.Foo(<>c.<>9__1_0 ?? (<>c.<>9__1_0 = new Action<D>(<>c.<>9.<M>b__1_0)));
        d.Foo(<>c.<>9__1_1 ?? (<>c.<>9__1_1 = new Action<D>(<>c.<>9.<M>b__1_1)));
    }
}
```

(the lambdas might also be written `static (dm1) => dm1.Bar (42,42)` with `static` explicitly asking for an error if the lambda closes over anything)

An EnC update of the body of a static lambda ends up creating additional static fields in the closure object and new static methods for the lambda bodies.

So to support these we really need the "add static field" hack from coreclr and "add static method" support

**Comment by lewing** [comment]
@lambdageek is this still targeting 6.0?

**Comment by lambdageek** [comment]
Emitting new lambdas is going to be 7.

Updating the bodies of existing lambdas might happen in 6 - there are some circumstances where `dotnet watch`+Roslyn emit a delta that puts an updated lambda body into a new method where in CI we see Roslyn update an existing lambda body.  We should be able to get `dotnet watch`+Roslyn to update existing lambdas, too.

**Comment by lambdageek** [comment]
Updating of existing lambda bodies that are generated from .razor pages will become supported once https://github.com/dotnet/roslyn/pull/55407 makes it into the dotnet SDK.

Adding new lambda bodies will be .net 7

**Comment by lambdageek** [comment]
Adding new lambdas (either static lambdas that don't capture anything, or new lambdas that capture `this` when there is already at least one lambda that captures `this`, or lambdas that capture locals (as long as there's already at least one lambda that captures the same locals) will be supported in the first preview of .NET 7 (Fixed in #63513).
